modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/dbf/index/ExpressionFilterVisitor.java

```diff
@@ -0,0 +1,395 @@
+        this.indexNodeMap = indexNodeMap;
+    }
+    
+    /**
+     * Determines if the filter and all its sub filters and expressions are supported.
+     */
+    public boolean fullySupports(Filter filter) {
+        return ExpressionFilterVisitor.capabilities.fullySupports(filter);
+    }
+    
+    /**
+     * Overrides AttributeImpl to avoid the costly 'parse' method.
+     */
+    class FastAttributeImpl extends AttributeImpl {
+        
+        public FastAttributeImpl(Object content, AttributeType type, Identifier id) {
+            super(content, type, id);
+        }
+        @Override
+        protected Object parse(Object value) throws IllegalArgumentException {
+            return value;
+        }
+    }
+    /**
+     * Helper argument class of one filter evaluation.
+     */
+    class FilterVisitorArgs implements ComplexPropertyAccessor, NodeVisitorArgs {
+        
+        public FilterVisitorArgs(Filter filter) {
+            
+            FilterAttributeExtractor attributeExtractor = new FilterAttributeExtractor();
+            filter.accept(attributeExtractor, null);
+            Set<PropertyName> propertyNameList = attributeExtractor.getPropertyNameSet();
+            
+            this.propertyNameArray = propertyNameList.toArray(new PropertyName[0]);
+            this.propertyArray = new Property[propertyNameList.size()];
+            this.filter = filter;
+        }
+        PropertyName[] propertyNameArray;
+        Property[] propertyArray;
+        Filter filter;
+        Map<Integer,Integer> result = new HashMap<Integer,Integer>();
+        List<Integer> buffer = new ArrayList<Integer>();
+        int evaluateCount;
+        boolean cancel;
+        
+        /**
+         * Returns the current filter being evaluated.
+         */
+        @Override
+        public Filter filter() {
+            return filter;
+        }
+        
+        /**
+         * Returns {@code true} if this job is cancelled.
+         */
+        @Override
+        public boolean isCanceled() {
+            return cancel;
+        }
+        
+        /**
+         * Indicates that task should be cancelled.
+         */
+        @Override
+        public void setCanceled(boolean cancel) {
+            this.cancel = cancel;
+        }
+        
+        /**
+         * Returns a shared buffer to efficiently visit nodes.
+         */
+        @Override
+        public byte[] sharedBuffer(int bufferLength) {
+            byte[] buffer = sharedBuffer;
+            
+            if (buffer==null || buffer.length<bufferLength) {
+                sharedBuffer = new byte[bufferLength];
+                buffer = sharedBuffer;
+            }
+            return buffer;
+        }
+        private byte[] sharedBuffer;
+        
+        /**
+         * Returns a single property of the complex object which matches the specified name.
+         */
+        @Override
+        public Property getProperty(String propertyName) {
```

Comment:
So if we are only using a propertyName here - and not an xpath/with/several/levels ... why do we need this getProperty method again?

We already have an interface (called SimpleFeature that can wrap around an propertyArray and perform property lookup.
```
Property property = feature.getProperty( propertyName ); // returns first match
for( Property attribute : feature.getProperties( propertyName) ){
   // handles all matches in case of duplicates
}
```

For more information see: http://docs.geotools.org/latest/javadocs/org/opengis/feature/ComplexAttribute.html